cmd/server/main.go

```diff
@@ -1,53 +1,148 @@
 package main
 
 import (
+	"bytes"
+	"context"
 	"fmt"
-	"net"
-	"time"
+	"io"
+	"log"
+	"net/http"
+	"runtime"
+	"strings"
 
+	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/onflow/flow-evm-gateway/api"
+	"github.com/onflow/flow-evm-gateway/indexer"
+	"github.com/onflow/flow-evm-gateway/storage"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog"
+)
+
+const (
+	accessURL = "access-001.devnet49.nodes.onflow.org:9000"
 )
 
 func main() {
-	server := api.NewRPCServer()
-	defer server.Stop()
+	store := storage.NewStore()
+	runServer(store)
+	runIndexer(store)
+
+	runtime.Goexit()
+}
+
+func apis(store *storage.Store) []rpc.API {
+	return []rpc.API{
+		{
+			Namespace: api.EthNamespace,
+			Service: &api.BlockChainAPI{
+				Store: store,
+			},
+		},
+	}
+}
+
+func runIndexer(store *storage.Store) {
+	ctx := context.Background()
 
-	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	chain, err := indexer.GetChain(ctx, accessURL)
 	if err != nil {
-		fmt.Println("can't listen:", err)
+		log.Fatalf("could not get chain: %v", err)
 	}
-	defer listener.Close()
-	go server.ServeListener(listener)
 
+	execClient, err := indexer.NewExecutionDataClient(accessURL, chain)
+	if err != nil {
+		log.Fatalf("could not create execution data client: %v", err)
+	}
+
+	sub, err := execClient.SubscribeEvents(ctx, flow.ZeroID, 0, indexer.EventFilter{
+		Contracts: []string{"A.7e60df042a9c0868.FlowToken"},
+	})
+	if err != nil {
+		log.Fatalf("could not subscribe to execution data: %v", err)
+	}
+
+	for {
```

Comment:
you should check in this loop that responses were received for all block heights, otherwise it's possible that a response was missed.

Reply:
How would that check look like? :thinking: 

To share some more context, we want to keep track of the 2 following service events (a.k.a core events):
```go
sub, err = execClient.SubscribeEvents(
	ctx,
	flow.ZeroID,
	height,
	indexer.EventFilter{
		EventTypes: []string{
			"flow.evm.BlockExecuted",
			"flow.evm.TransactionExecuted",
		},
	},
)
```
and build some sort of state around the payloads of these 2 events.

These 2 events will start being emitted from a certain block height and onwards. That block height would most likely be the height in which the `EVM` contract (and associated code) is deployed/introduced on `testnet` & `mainnet`. For this case we can have a utility script for back-filling.

My assumption here is that `SubscribeEvents` will publish a response for every block that matches these `EventFilter` criteria. Of course, not every block will contain such events. In the context of this Flow EVM Gateway, the latest block height is the block height which contained a `flow.evm.BlockExecuted`, we are not interested in regular Flow blocks.

Could you elaborate on which case it is possible to miss a response which matched the `EventFilter` criteria?

Reply:
The `SubscribeEvents` endpoint has a `heartbeatInterval` request parameter:
https://github.com/onflow/flow/blob/b7ed2c3e452b5e44561399afc6b5e868e8957117/protobuf/flow/executiondata/executiondata.proto#L140-L148

This is used to set how often the backend will return a response message. If you set it to 1, it will return a response for every block. If the block has no events, the list is empty. If you check the block height from each message increases sequentially, you can verify that you received all messages, and thus all blocks were searched.

The API is currently missing an important sequence number field, which would let you get the same guarantee with larger `heartbeatIntervals` without receiving a response for every block.

Given missing any events may corrupt your index, I'd recommend using the 1 block heartbeat for now.

You'll also want to add some reconnect handling similar to what's implement here:
https://github.com/onflow/flow-go-sdk/pull/417/files#diff-5bf26624cf4a959cb8bcbe675b61ae59dbdcd976605d3df3c5de93aa30e0a32bR62

Reply:
Thanks for the detailed explanation, I was not aware of the `heartbeatInterval` request parameter.
I have added this parameter and the relevant check in https://github.com/onflow/flow-evm-gateway/pull/11/commits/2b36693bad4024e94cb511858a5f156cb6afd34f.
I will also follow up with a PR on upstream (https://github.com/peterargue/execdata-client/blob/main/client/client.go#L135-L140). I believe it can be useful.